File: pipeline/scripts/geocode_stores.py
#!/usr/bin/env python3

import json
import time
import urllib.parse
import urllib.request
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
OPENPILOT_CARS_PATH = SCRIPT_DIR.parent / "data" / "openpilot_cars.json"
STORE_COORDS_PATH = SCRIPT_DIR.parent / "data" / "store-coords.json"
USER_AGENT = "openpilot-car-finder/1.0 (geocoding carmax stores)"
REQUEST_DELAY_SECONDS = 1.1


def load_json(path: Path) -> dict:
  return json.loads(path.read_text())


def save_json(path: Path, value: dict) -> None:
  path.write_text(json.dumps(value, indent=2) + "\n")


def collect_stores_by_id(cars_data: dict) -> dict[str, dict]:
  stores_by_id: dict[str, dict] = {}
  for entry in cars_data.get("entries", []):
    for available_year in entry.get("available_years", []):
      car = available_year.get("car")
      if car is None:
        continue

      store_id = str(car["storeId"])
      if store_id in stores_by_id:
        continue

      stores_by_id[store_id] = {
        "id": store_id,
        "name": car.get("storeName", ""),
        "city": car.get("storeCity", ""),
        "state": car.get("state", ""),
      }
  return stores_by_id


def build_search_url(query: str) -> str:
  encoded_query = urllib.parse.quote(query, safe="")
  return f"https://nominatim.openstreetmap.org/search?q={encoded_query}&format=json&limit=1"


def fetch_first_geocode_result(query: str) -> dict | None:
  request = urllib.request.Request(
    build_search_url(query),
    headers={"User-Agent": USER_AGENT},
  )
  with urllib.request.urlopen(request) as response:
    if response.status != 200:
      raise RuntimeError(f"HTTP {response.status} for query: {query}")
    results = json.load(response)
  return results[0] if results else None


def to_store_coords(geocode_result: dict) -> dict:
  return {
    "lat": float(geocode_result["lat"]),
    "lng": float(geocode_result["lon"]),
  }


def geocode_store(store: dict) -> dict | None:
  primary_query = f"CarMax {store['city']}, {store['state']}"
  primary_result = fetch_first_geocode_result(primary_query)
  if primary_result is not None:
    return to_store_coords(primary_result)

  time.sleep(REQUEST_DELAY_SECONDS)
  fallback_query = f"{store['city']}, {store['state']}"
  fallback_result = fetch_first_geocode_result(fallback_query)
  return to_store_coords(fallback_result) if fallback_result is not None else None


def sort_coords_by_store_id(coords: dict) -> dict:
  return dict(sorted(coords.items(), key=lambda item: int(item[0])))


def main() -> int:
  cars_data = load_json(OPENPILOT_CARS_PATH)
  stores = collect_stores_by_id(cars_data)
  print(f"Found {len(stores)} unique stores. Starting geocoding...")

  coords: dict[str, dict] = {}
  failed_stores: list[dict] = []
  done_count = 0

  for store in stores.values():
    try:
      geocoded = geocode_store(store)
      if geocoded is not None:
        coords[store["id"]] = geocoded
        done_count += 1
        if done_count % 20 == 0:
          print(f"  {done_count}/{len(stores)} done...")
      else:
        failed_stores.append(store)
        print(
          f'FAILED: storeId={store["id"]} name="{store["name"]}" '
          f'city="{store["city"]}" state="{store["state"]}"'
        )
    except Exception as error:  # noqa: BLE001 - keep full context in CLI output
      failed_stores.append(store)
      print(
        f'ERROR: storeId={store["id"]} name="{store["name"]}" '
        f'city="{store["city"]}" state="{store["state"]}" -> {error}'
      )

    time.sleep(REQUEST_DELAY_SECONDS)

  save_json(STORE_COORDS_PATH, sort_coords_by_store_id(coords))
  print(f"\nDone! {done_count} stores geocoded, {len(failed_stores)} failed.")

  if failed_stores:
    print(f"Failed stores: {failed_stores}")

  return 1 if failed_stores else 0


if __name__ == "__main__":
  raise SystemExit(main())

File: pipeline/scripts/geocode-stores.mjs
/**
 * One-time script: geocodes all CarMax stores in openpilot_cars.json
 * via Nominatim (OpenStreetMap) and writes pipeline/data/store-coords.json.
 *
 * Run: node scripts/geocode-stores.mjs
 * Takes ~4 min due to Nominatim's 1 req/sec rate limit.
 */

import { readFileSync, writeFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const USER_AGENT = "openpilot-car-finder/1.0 (geocoding carmax stores)";
const carsData = JSON.parse(readFileSync(join(__dirname, "../data/openpilot_cars.json"), "utf-8"));

// Collect unique stores
const stores = new Map();
for (const entry of carsData.entries) {
  for (const availableYear of entry.available_years) {
    if (availableYear.car && !stores.has(availableYear.car.storeId)) {
      stores.set(availableYear.car.storeId, {
        id: availableYear.car.storeId,
        name: availableYear.car.storeName,
        city: availableYear.car.storeCity,
        state: availableYear.car.state,
      });
    }
  }
}

console.log(`Found ${stores.size} unique stores. Starting geocoding...`);

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function buildSearchUrl(query) {
  return `https://nominatim.openstreetmap.org/search?q=${encodeURIComponent(query)}&format=json&limit=1`;
}

async function fetchFirstGeocodeResult(query) {
  const response = await fetch(buildSearchUrl(query), {
    headers: { "User-Agent": USER_AGENT },
  });
  const results = await response.json();
  return results[0] ?? null;
}

function toStoreCoords(geocodeResult) {
  return {
    lat: parseFloat(geocodeResult.lat),
    lng: parseFloat(geocodeResult.lon),
  };
}

const coords = {};
let done = 0;
let failed = [];

for (const store of stores.values()) {
  try {
    const primaryResult = await fetchFirstGeocodeResult(`CarMax ${store.city}, ${store.state}`);

    if (primaryResult) {
      coords[store.id] = toStoreCoords(primaryResult);
      done++;
      if (done % 20 === 0) console.log(`  ${done}/${stores.size} done...`);
    } else {
      // Fallback: try just city + state without "CarMax"
      await sleep(1100);
      const fallbackResult = await fetchFirstGeocodeResult(`${store.city}, ${store.state}`);
      if (fallbackResult) {
        coords[store.id] = toStoreCoords(fallbackResult);
        done++;
      } else {
        console.warn(`  FAILED: storeId=${store.id} name="${store.name}" city="${store.city}" state="${store.state}"`);
        failed.push(store);
      }
    }
  } catch (err) {
    console.error(`  ERROR: storeId=${store.id}:`, err.message);
    failed.push(store);
  }

  // Respect Nominatim rate limit: max 1 request/second
  await sleep(1100);
}

writeFileSync(join(__dirname, "../data/store-coords.json"), JSON.stringify(coords, null, 2));
console.log(`\nDone! ${done} stores geocoded, ${failed.length} failed.`);
if (failed.length > 0) {
  console.log("Failed stores:", failed);
}
